### ISSUE TITLE
migrate_to_binary break downstream tests

### ISSUE THREAD

**Issue by MidnightWonderer** [RailsEventStore/rails_event_store#329]
Hi there,  
This break my test:  
https://github.com/RailsEventStore/rails_event_store/blob/552dfb4/ruby_event_store/lib/ruby_event_store/spec/event_repository_lint.rb#L980  
I wanted to test for binary compatibility but I have nothing to migrate.  
I think the operation should be optional otherwise I'll end up with  
```ruby
let(:migrate_to_binary) { nil }
```
with no apparent reason.  
At least maybe a rescue and ignore?

**Comment by paneq** [comment]
When in doubt check what other repos do:

* https://github.com/RailsEventStore/rails_event_store/blob/bb73cbce7c5e264aec16fe06d5470e89d504cd8c/rails_event_store_active_record/spec/event_repository_spec.rb#L265
* https://github.com/RailsEventStore/rails_event_store/blob/bb73cbce7c5e264aec16fe06d5470e89d504cd8c/ruby_event_store/spec/in_memory_repository_spec.rb#L66

```ruby
    def migrate_to_binary
    end
```

is perfectly fine.

**Comment by MidnightWonderer** [comment]
Hi @paneq  
Just to make the point clear: I am not in doubt.  
I have a strong opinion that `migrate_to_binary` should not be there in the `ruby_event_store` gem.  
  
This violate the direction of dependency.  
`rails_event_store_active_record` depends on `ruby_event_store` hence `ruby_event_store` should not know anything about the "migration".  

It is unfortunate that `rails_event_store_active_record` previously has an incorrect migration but this is not the place to workaround the issue.  
  
However, you are the maintainer of the gem so I just don't strongly oppose it at first. I only ask it won't force new implementations to conform to the interface in question.

**Comment by paneq** [comment]
@MidnightWonderer 

That is the unfortunate limitation of using Template Method pattern.

Ideally I would perform the migration before running that single test. Unfortunately I don't know how to easily achieve that with rspec shared examples. If you know, let me know.


That's probably one of the places where RSpec DSL is poor substitute for including modules and/or inheritance in Ruby and Minitest.

```ruby
module TestBinary
  def test_binary
    ...
  end
end

class RepoTest
  include TestBinary

  def test_binary
    run_migration
    super
  end
end
```

**Comment by MidnightWonderer** [comment]
Glad to hear your opinion.  
  
Just for this specific case. My concern is only about the interface enforcement.  
The hook can be there but with  
```ruby
require 'active_support/core_ext/object/try'
...
try(:migrate_to_binary)
```
instead of just
```ruby
migrate_to_binary
```
  
In more general about RSpec shared example, I have no better experience than you but just some idea form the OOP school.  
It seems that we can supply external parameters to an RSpec shared example.  
As seen here:  
```ruby
RSpec.shared_examples "some example" do |parameter|
  \# Same behavior is triggered also with either `def something; 'some value'; end`
  \# or `define_method(:something) { 'some value' }`
  let(:something) { parameter }
  it "uses the given parameter" do
    expect(something).to eq(parameter)
  end
end

RSpec.describe SomeClass do
  include_examples "some example", "parameter1"
  include_examples "some example", "parameter2"
end
```
reference:  
https://relishapp.com/rspec/rspec-core/v/3-7/docs/example-groups/shared-examples  
To me this
```ruby
include_examples "some example", "parameter2"
```
look just like a class construction and a method invocation
```ruby
spec_example = SomeExample.new('parameter2')
spec_example.run
```
So all the DI tricks you use throughout the code base seems to work with shared examples too. (I guess, haven't tried)  
From there you can supply a lambda to migrate the database.  
  
To sum it all up I am thinking along the line of:  
```ruby
RSpec.shared_examples :binary_test do |before_test: ->() {}|
  specify 'can store arbitrary binary data' do
    before_test.call
    binary = "\xB0"
    expect(binary.valid_encoding?).to eq(false)
    binary.force_encoding("binary")
    expect(binary.valid_encoding?).to eq(true)

    repository.append_to_stream(
      event = SRecord.new(data: binary, metadata: binary),
      RubyEventStore::Stream.new('stream'),
      RubyEventStore::ExpectedVersion.none
    )
  end
end
```
I'm not sure if this work but this would break the API for the spec.  
  
For the current API I will simply mention this thread.
https://github.com/the-cave/ruby_event_store_event_id_mapper/blob/7bc68f6/spec/repository_wrapper_spec.rb#L26  
If you break the API for goods I am willing to follow.

**Comment by paneq** [comment]
> `try(:migrate_to_binary)`

That's what I would rather avoid.

> include_examples "some example", "parameter2"

I've been looking at exactly the same thing before answering you.

I imagined it would require some implementations to pass empty proc or a null object (similar to what you showed). And to me that's really no different than defining an empty method. So in the end I decided not to go this way (for now).

Now we can debate whether to use more general naming such as `before_test` or more specific such as `migrate_to_binary`. I prefer to start with a specific name and generalize once there are more _clients_ of the method which seem to use it in a more generalized way.

> Glad to hear your opinion.

glad to hear yours :)
